xapian-core/weight/tfidfweight.cc

```diff
@@ -49,10 +50,31 @@ TfIdfWeight::TfIdfWeight(const std::string &normals)
     need_stat(WDF_MAX);
 }
 
+TfIdfWeight::TfIdfWeight(const std::string &normals, double slope, double delta)
+    : normalizations(normals), param_slope(slope), param_delta(delta)
+{
+    if (normalizations.length() != 3 ||
+	!strchr("nbslP", normalizations[0]) ||
+	!strchr("ntpfsP", normalizations[1]) ||
+	!strchr("nP", normalizations[2]))
+	throw Xapian::InvalidArgumentError("Normalization string is invalid");
+    if (normalizations[1] != 'n') {
+	need_stat(TERMFREQ);
+	need_stat(COLLECTION_SIZE);
+    }
+    need_stat(WDF);
+    need_stat(WDF_MAX);
+    if (param_slope != 0 || param_delta != 0) {
+	need_stat(AVERAGE_LENGTH);
+	need_stat(DOC_LENGTH);
+	need_stat(WQF);
+    }
+}
+
 TfIdfWeight *
 TfIdfWeight::clone() const
 {
-    return new TfIdfWeight(normalizations);
+	return new TfIdfWeight(normalizations, param_slope, param_delta);
 }
```

Comment:
Sorry, I was planning to make changes at one go after getting your input on some comments below. Anyways, fixing these formatting issues now.